### PR TITLE
Add `SetRequired` and `SetOptional` types

### PR DESCRIPTION
**Pull request by crhistianramirez**
~Now that I'm looking at it maybe SomeRequired and SomeOptional would be better names? Couples the two, as it should and expresses intent pretty well.~

Edit: changed to SetRequired and SetOptional per feedback from sindresorhus

### CODE REVIEW

**Comment by crhistianramirez** [comment]
Now that I'm looking at it maybe `SomeRequired` and `SomeOptional` would be a better name?

**Comment by sindresorhus** [comment]
> maybe SomeRequired and SomeOptional would be better names?

👍 

It's really too bad that the built-in `Required` type is not named `RequiredAll` or `AllRequired`...

File: source/optionalize.d.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/**
+The sister of `Require` type. Allows you to define one or more keys for a type that should be optional.
```
Comment (sindresorhus) [inline source/optionalize.d.ts]: The sentences here should be in reverse order.

File: source/optionalize.d.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+Use-case: You want to define a single model where the only thing that changes is
+whether or not some of the properties are optional
```
Comment (sindresorhus) [inline source/optionalize.d.ts]: Don't hard-wrap

File: source/optionalize.d.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  a: number;
+  b?: string;
+  c: boolean;
+}
+
+type SomeOptional = Optionalize<Foo, 'b' | 'c'>;
+// type SomeOptional = {
+//   a: number;
+//   b?: string; // was already optional, still is
+//   c?: boolean; // is now optional
```
Comment (sindresorhus) [inline source/optionalize.d.ts]: Use tab-indentation

File: source/optionalize.d.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+// }
+```
+*/
+
+export type Optionalize<BaseType, Keys extends keyof BaseType = keyof BaseType> =
+Pick<BaseType, Exclude<keyof BaseType, Keys>> &
+  Partial<Pick<BaseType, Keys>> extends infer InferredType
+  ? { [Property in keyof InferredType]: InferredType[Property] }
+  : never;
```
Comment (sindresorhus) [inline source/optionalize.d.ts]: > At the bottom, explain how the type works. Some types can be quite advanced and hard to understand. We can use this opportunity to teach users. - https://github.com/sindresorhus/type-fest/blob/master/.github/contributing.md

**Comment by sindresorhus** [comment]
You need to add the types to the readme.

File: index.d.ts [hunk 5]
```diff
@@ -6,9 +6,11 @@ export {Except} from './source/except';
 export {Mutable} from './source/mutable';
 export {Merge} from './source/merge';
 export {MergeExclusive} from './source/merge-exclusive';
+export {Require} from './source/require';
```
Comment (sindresorhus) [inline index.d.ts]: This should be moved to line 12.

File: source/optionalize.d.ts (hunk 1 above)
Comment (sindresorhus) [inline source/optionalize.d.ts]: > Allows you to define one or more keys for a type that should be optional.

This is not entirely correct as it creates a new type. It doesn't modify the existing type.

**Comment by crhistianramirez** [comment]
@sindresorhus thank you for the feedback I went ahead and resolved each line item. I'd appreciate additional feeedback you might have :)

Also did you prefer that I squash all the commits to clean it up? Wasn't sure, the readme didn't mention it.

**Comment by sindresorhus** [comment]
Looks good now. Thanks :)

**Pull request merged**